File: Weijia/740. Delete and Earn.py
class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        l=[0]*(max(nums))
        
        for i in range(1,max(nums)+1):
            if i in nums:
                l[i-1]=i*nums.count(i)
        print(l)
        if len(l)<=2:
            return max(l)
        else:
            m=[l[0],max(l[0],l[1])]
            for n in range(2,len(l)):
                m=[m[-1],max(m[-1],m[-2]+l[n])]
                print(m)
            return m[-1]

File: solution_zzj/740._Delete_and_Earn.cc
class Solution {
public:
    int deleteAndEarn(vector<int>& nums) {
        int mx = *max_element(nums.begin(),nums.end());
        vector<int> vec(mx, 0);
        for (auto num: nums) {
            vec[num-1] += num;
        }
        if (vec.size()==1) return vec[0];
        vector<int> ans = {vec[0], max(vec[0], vec[1])};
        for (int i=2;i<vec.size();i++) {
            ans.push_back(max(ans[i-2]+vec[i], ans[i-1]));
        }
        return ans[vec.size()-1];
    }
};
